File: sdk/src/merkle.ts
// The keccak Merkle tree over one pool key's per-member share commitments.
// `finalizeEpoch` stores its root (`poolShareRoots[eid][j]`) and
// `submitPartialDecryption` proves the member's leaf against it; this module
// mirrors `circuits/common/merkle.go` and the tree the contract folds so the
// SDK can build and check that path. Tagged encoding, so no leaf can be
// replayed as an internal node and an absent member is distinguishable from
// any point:
//
//   leaf[p-1] = keccak256(0x00 ‖ D_p.x ‖ D_p.y)      for committee member p
//   leaf[i]   = keccak256("davinci-dkg:merkle-empty:v1")   for i ≥ committeeSize
//   node      = keccak256(0x01 ‖ left ‖ right)
//
// Coordinates are 32-byte big-endian words (the on-chain RTE form); the tree
// is a fixed `MERKLE_DEPTH` levels over `MAX_N` leaves, indexed by
// participant index − 1.

import { keccak256, toHex, type Hex } from 'viem';
import { MAX_N, MERKLE_DEPTH } from './sizes.js';
import { bytes32, type TranscriptPoint } from './transcript.js';

/** Leaf of a committee slot that does not exist (`i ≥ committeeSize`). */
export const MERKLE_EMPTY_LEAF: Hex = keccak256(toHex('davinci-dkg:merkle-empty:v1'));

const LEAF_TAG = '00';
const NODE_TAG = '01';

/** `keccak256(0x00 ‖ D.x ‖ D.y)`: one member's share commitment as its leaf. */
export function shareCommitmentLeaf(commitment: TranscriptPoint): Hex {
  return keccak256(('0x' + LEAF_TAG + bytes32(commitment.x).slice(2) + bytes32(commitment.y).slice(2)) as Hex);
}

/** `keccak256(0x01 ‖ left ‖ right)`: two children into their parent. */
export function merkleNode(left: Hex, right: Hex): Hex {
  return keccak256(('0x' + NODE_TAG + left.slice(2) + right.slice(2)) as Hex);
}

/**
 * Lay one pool key's share commitments out as the `MAX_N` leaves: slot `i`
 * (committee member `i + 1`) is `shareCommitmentLeaf(D_{i+1})` for
 * `i < committeeSize` and the empty leaf beyond. `shareCommitments` is in
 * committee order and may be the full padded row from the finalize
 * transcript; only the first `committeeSize` entries are read.
 */
export function shareCommitmentLeaves(
  shareCommitments: readonly TranscriptPoint[],
  committeeSize: number,
): Hex[] {
  if (!Number.isInteger(committeeSize) || committeeSize < 0 || committeeSize > MAX_N) {
    throw new Error(`merkle: committee size ${committeeSize} out of range [0, ${MAX_N}]`);
  }
  if (shareCommitments.length < committeeSize) {
    throw new Error(`merkle: got ${shareCommitments.length} share commitments for a committee of ${committeeSize}`);
  }
  const leaves: Hex[] = new Array<Hex>(MAX_N).fill(MERKLE_EMPTY_LEAF);
  for (let i = 0; i < committeeSize; i++) leaves[i] = shareCommitmentLeaf(shareCommitments[i]);
  return leaves;
}

function merkleLevel(nodes: readonly Hex[]): Hex[] {
  const next: Hex[] = [];
  for (let i = 0; i < nodes.length; i += 2) next.push(merkleNode(nodes[i], nodes[i + 1]));
  return next;
}

function requireLeaves(leaves: readonly Hex[]): void {
  if (leaves.length !== MAX_N) throw new Error(`merkle: expected ${MAX_N} leaves, got ${leaves.length}`);
}

/** Fold the `MAX_N` leaves into the root over `MERKLE_DEPTH` levels. */
export function merkleRoot(leaves: readonly Hex[]): Hex {
  requireLeaves(leaves);
  let level: Hex[] = [...leaves];
  for (let depth = 0; depth < MERKLE_DEPTH; depth++) level = merkleLevel(level);
  return level[0];
}

/**
 * The siblings of leaf `index` bottom-up — the `shareProof` argument of
 * `submitPartialDecryption` for participant `index + 1`.
 */
export function merklePath(leaves: readonly Hex[], index: number): Hex[] {
  requireLeaves(leaves);
  if (!Number.isInteger(index) || index < 0 || index >= MAX_N) {
    throw new Error(`merkle: leaf index ${index} out of range [0, ${MAX_N})`);
  }
  const path: Hex[] = [];
  let level: Hex[] = [...leaves];
  let cursor = index;
  for (let depth = 0; depth < MERKLE_DEPTH; depth++) {
    path.push(level[cursor ^ 1]);
    cursor >>= 1;
    level = merkleLevel(level);
  }
  return path;
}

/**
 * Recompute the root a `MERKLE_DEPTH`-long path claims for `leaf` at `index`
 * — exactly the fold `submitPartialDecryption` runs before comparing with
 * `poolShareRoots`.
 */
export function merkleRootFromPath(leaf: Hex, index: number, path: readonly Hex[]): Hex {
  if (path.length !== MERKLE_DEPTH) throw new Error(`merkle: expected a ${MERKLE_DEPTH}-long path, got ${path.length}`);
  if (!Number.isInteger(index) || index < 0 || index >= MAX_N) {
    throw new Error(`merkle: leaf index ${index} out of range [0, ${MAX_N})`);
  }
  let node = leaf;
  let cursor = index;
  for (const sibling of path) {
    node = cursor & 1 ? merkleNode(sibling, node) : merkleNode(node, sibling);
    cursor >>= 1;
  }
  return node;
}

/** True when `path` proves `leaf` at `index` under `root`. */
export function verifyMerklePath(root: Hex, leaf: Hex, index: number, path: readonly Hex[]): boolean {
  try {
    return merkleRootFromPath(leaf, index, path).toLowerCase() === root.toLowerCase();
  } catch {
    return false;
  }
}

/**
 * The `shareProof` of committee member `participantIndex` (1-based) under one
 * pool key, from that key's share commitments in committee order (e.g.
 * `FinalizeTranscript.shareCommitments[j]`), together with the leaf and the
 * root the contract must hold.
 */
export function shareProof(
  shareCommitments: readonly TranscriptPoint[],
  committeeSize: number,
  participantIndex: number,
): { leaf: Hex; path: Hex[]; root: Hex } {
  if (!Number.isInteger(participantIndex) || participantIndex < 1 || participantIndex > committeeSize) {
    throw new Error(`merkle: participant index ${participantIndex} out of range [1, ${committeeSize}]`);
  }
  const leaves = shareCommitmentLeaves(shareCommitments, committeeSize);
  const index = participantIndex - 1;
  return { leaf: leaves[index], path: merklePath(leaves, index), root: merkleRoot(leaves) };
}

File: circuits/common/merkle.go
package common

import (
	"fmt"

	ethcrypto "github.com/ethereum/go-ethereum/crypto"
	"github.com/vocdoni/davinci-dkg/types"
)

// Compile-time guard: the share-commitment tree has exactly MaxN leaves, so
// MerkleDepth must stay log2(MaxN). Either term underflows uint otherwise.
const _ = uint(1<<MerkleDepth-MaxN) + uint(MaxN-1<<MerkleDepth)

// The keccak Merkle tree over one pool key's per-member share commitments:
// activatePoolKey stores its root and submitPartialDecryption proves the
// member's leaf against it. Keccak (not Poseidon) because only the contract
// ever recomputes a path. See docs/pool-keys.md.
//
// Tagged encoding, so no leaf can be replayed as an internal node (or the
// other way round) and an absent member is distinguishable from any point:
//
//	leaf[p-1] = keccak256(0x00 ‖ D_p.x ‖ D_p.y)      for committee member p
//	leaf[i]   = EmptyLeaf                           for i ≥ committeeSize
//	node      = keccak256(0x01 ‖ left ‖ right)
//
// Coordinates are 32-byte big-endian words; the tree is a fixed MerkleDepth
// levels over MaxN leaves, indexed by participant index − 1.

const (
	merkleLeafTag byte = 0x00
	merkleNodeTag byte = 0x01
)

// EmptyLeaf fills every leaf beyond the committee. It is a fixed domain
// hash rather than bytes32(0) so an empty slot is never confused with a
// leaf whose hash happens to be zero-like, and the constant is what the
// contract must mirror.
var EmptyLeaf = [32]byte(ethcrypto.Keccak256([]byte("davinci-dkg:merkle-empty:v1")))

// ShareCommitmentLeaf encodes one member's share commitment as its leaf.
func ShareCommitmentLeaf(commitment types.CurvePoint) ([32]byte, error) {
	if err := commitment.Validate(); err != nil {
		return [32]byte{}, err
	}
	packed := make([]byte, 0, 65)
	packed = append(packed, merkleLeafTag)
	packed = append(packed, commitment.X.FillBytes(make([]byte, 32))...)
	packed = append(packed, commitment.Y.FillBytes(make([]byte, 32))...)
	return [32]byte(ethcrypto.Keccak256(packed)), nil
}

// MerkleNode hashes two children into their parent.
func MerkleNode(left, right [32]byte) [32]byte {
	packed := make([]byte, 0, 65)
	packed = append(packed, merkleNodeTag)
	packed = append(packed, left[:]...)
	packed = append(packed, right[:]...)
	return [32]byte(ethcrypto.Keccak256(packed))
}

// ShareCommitmentLeaves lays out one pool key's share commitments as the MaxN
// tree leaves: leaf[p-1] = ShareCommitmentLeaf(D_p) for every listed
// participant p, EmptyLeaf for every other slot. Callers pass the whole
// committee (indexes 1..committeeSize), because the activation transcript
// carries a D_p for every member, contributing or not.
func ShareCommitmentLeaves(
	participantIndexes []uint16,
	shareCommitments []types.CurvePoint,
) ([MaxN][32]byte, error) {
	var leaves [MaxN][32]byte
	for i := range leaves {
		leaves[i] = EmptyLeaf
	}
	if len(participantIndexes) != len(shareCommitments) {
		return leaves, fmt.Errorf(
			"got %d participant indexes and %d share commitments",
			len(participantIndexes),
			len(shareCommitments),
		)
	}
	seen := make(map[uint16]struct{}, len(participantIndexes))
	for i, index := range participantIndexes {
		if index == 0 || int(index) > MaxN {
			return leaves, fmt.Errorf("participant index %d out of range [1, %d]", index, MaxN)
		}
		if _, duplicate := seen[index]; duplicate {
			return leaves, fmt.Errorf("duplicate participant index %d", index)
		}
		seen[index] = struct{}{}
		leaf, err := ShareCommitmentLeaf(shareCommitments[i])
		if err != nil {
			return leaves, fmt.Errorf("share commitment %d: %w", i, err)
		}
		leaves[index-1] = leaf
	}
	return leaves, nil
}

// MerkleRoot folds the leaves into the root over MerkleDepth levels of
// MerkleNode.
func MerkleRoot(leaves [MaxN][32]byte) [32]byte {
	level := make([][32]byte, MaxN)
	copy(level, leaves[:])
	for range MerkleDepth {
		level = merkleLevel(level)
	}
	return level[0]
}

// MerklePath returns the siblings of leaf `index` bottom-up: the proof
// submitPartialDecryption takes alongside the partial.
func MerklePath(leaves [MaxN][32]byte, index int) ([MerkleDepth][32]byte, error) {
	var path [MerkleDepth][32]byte
	if index < 0 || index >= MaxN {
		return path, fmt.Errorf("leaf index %d out of range [0, %d)", index, MaxN)
	}
	level := make([][32]byte, MaxN)
	copy(level, leaves[:])
	for depth := range MerkleDepth {
		path[depth] = level[index^1]
		index >>= 1
		level = merkleLevel(level)
	}
	return path, nil
}

func merkleLevel(nodes [][32]byte) [][32]byte {
	next := make([][32]byte, len(nodes)/2)
	for i := range next {
		next[i] = MerkleNode(nodes[2*i], nodes[2*i+1])
	}
	return next
}
